File: 22_专题/离线查询/根号分治/块状链表/BlockChain.go
// https://www.acwing.com/blog/content/28060/
package main

import (
	"fmt"
	"math"
)

func main() {
	test()
}

type IBlock[S, U, V any] interface {
	// 分裂整块，block1包含前k个元素，block2包含后面的元素.分裂后，可以回收原块内存.
	Split(k int32) (block1 IBlock[S, U, V], block2 IBlock[S, U, V])
	// 合并两个块，合并后，可以回收原块内存.
	Merge(other IBlock[S, U, V]) IBlock[S, U, V]
	// 在index下标之前插入元素.
	InserBefore(index int32, e V)
	Delete(index int32)
	Get(index int32) V
	Reverse()

	FullyQuery(sum S)
	PartialQuery(index int32, sum S)
	FullyUpdate(lazy U)
	PartialUpdate(index int32, lazy U)
	BeforePartialQuery()
	AfterPartialUpdate()
}

// 块状链表.
type BlockChain[S, U, V any] struct {
	Head, Tail *linkedNode[IBlock[S, U, V]]
	b, size    int32
}

// blockSize: 2 * (int32(math.Sqrt(float64(n))) + 1)
func NewBlockChain[S, U, V any](n int32, blockSupplier func(start, end int32) IBlock[S, U, V], blockSize int32) *BlockChain[S, U, V] {
	if blockSize == -1 {
		blockSize = 2 * (int32(math.Sqrt(float64(n))) + 1)
	}
	res := &BlockChain[S, U, V]{
		b:    blockSize,
		size: n,
		Head: &linkedNode[IBlock[S, U, V]]{},
		Tail: &linkedNode[IBlock[S, U, V]]{},
	}
	linkNode(res.Head, res.Tail)
	for start := int32(0); start < n; start += blockSize {
		end := start + blockSize
		if end > n {
			end = n
		}
		block := blockSupplier(start, end)
		node := &linkedNode[IBlock[S, U, V]]{}
		node.data = block
		node.size = end - start
		linkNode(res.Tail.prev, node)
		linkNode(node, res.Tail)
	}
	return res
}

func (bc *BlockChain[S, U, V]) _new2(b int32, supplier func() IBlock[S, U, V]) *BlockChain[S, U, V] {
	res := &BlockChain[S, U, V]{
		b:    b,
		Head: &linkedNode[IBlock[S, U, V]]{},
		Tail: &linkedNode[IBlock[S, U, V]]{},
	}
	block := supplier()
	node := &linkedNode[IBlock[S, U, V]]{}
	node.data = block
	res.b = b
	linkNode(res.Tail.prev, node)
	linkNode(node, res.Tail)
	return res
}

func (bc *BlockChain[S, U, V]) _new3(b int32, begin, end *linkedNode[IBlock[S, U, V]]) *BlockChain[S, U, V] {
	res := &BlockChain[S, U, V]{
		b:    b,
		Head: &linkedNode[IBlock[S, U, V]]{},
		Tail: &linkedNode[IBlock[S, U, V]]{},
	}
	linkNode(res.Head, begin)
	linkNode(end, res.Tail)
	res._maintain()
	return res
}

func (bc *BlockChain[S, U, V]) Get(index int32) V {
	if index < 0 {
		index += bc.size
	}
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		if node.size <= index {
			index -= node.size
			continue
		}
		node.data.BeforePartialQuery()
		return node.data.Get(index)
	}
	panic("Index out of bounds")
}

func (bc *BlockChain[S, U, V]) PrefixSize(block IBlock[S, U, V], include bool) int32 {
	res := int32(0)
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		if node.data == block {
			if include {
				res += node.size
			}
			break
		}
		res += node.size
	}
	return res
}

func (bc *BlockChain[S, U, V]) Split(k int32, blockSupplier func() IBlock[S, U, V]) (first, second *BlockChain[S, U, V]) {
	k++
	if k == 1 {
		return bc._new2(bc.b, blockSupplier), bc
	}
	if k > bc.size {
		return bc, bc._new2(bc.b, blockSupplier)
	}
	head := bc._splitKth(k)
	end := bc.Tail.prev
	linkNode(head.prev, bc.Tail)
	b := bc._new3(bc.b, head, end)
	bc._maintain()
	return bc, b
}

func (bc *BlockChain[S, U, V]) MergeDestructively(other *BlockChain[S, U, V]) *BlockChain[S, U, V] {
	linkNode(bc.Tail.prev, other.Head.next)
	bc.Tail = other.Tail
	bc.size += other.size
	return bc
}

func (bc *BlockChain[S, U, V]) InsertBefore(index int32, e V) {
	if index < 0 {
		index += bc.size
	}
	if index < 0 {
		index = 0
	}
	if index > bc.size {
		index = bc.size
	}
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		if node.size < index {
			index -= node.size
			continue
		}
		node.data.InserBefore(index, e)
		node.size++
		break
	}
	bc._maintain()
}

func (bc *BlockChain[S, U, V]) Delete(index int32) {
	if index < 0 {
		index += bc.size
	}
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		if node.size <= index {
			index -= node.size
			continue
		}
		node.data.Delete(index)
		node.size--
		break
	}
	bc._maintain()
}

func (bc *BlockChain[S, U, V]) Update(start, end int32, update U) {
	if start < 0 {
		start += bc.size
	}
	if end > bc.size {
		end = bc.size
	}
	if start >= end {
		return
	}
	end--
	offset := int32(0)
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		left := offset
		right := offset + node.size - 1
		offset += node.size
		if bc._enter(start, end, left, right) {
			node.data.FullyUpdate(update)
		} else if bc._leave(start, end, left, right) {
			continue
		} else {
			for i := max32(left, start); i <= min32(right, end); i++ {
				node.data.PartialUpdate(i-left, update)
			}
			node.data.AfterPartialUpdate()
		}
	}
}

func (bc *BlockChain[S, U, V]) Query(start, end int32, sum S) {
	if start < 0 {
		start += bc.size
	}
	if end > bc.size {
		end = bc.size
	}
	if start >= end {
		return
	}
	end--
	offset := int32(0)
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		left := offset
		right := offset + node.size - 1
		offset += node.size
		if bc._enter(start, end, left, right) {
			node.data.FullyQuery(sum)
		} else if bc._leave(start, end, left, right) {
			continue
		} else {
			node.data.BeforePartialQuery()
			for i := max32(left, start); i <= min32(right, end); i++ {
				node.data.PartialQuery(i-left, sum)
			}
		}
	}
}

func (bc *BlockChain[S, U, V]) RotateLeft(k int32) {
	if k < 0 {
		k += bc.size
	}
	if k >= bc.size {
		k %= bc.size
	}
	if k == 0 {
		return
	}
	k++
	node := bc._splitKth(k)
	h1 := bc.Head.next
	e1 := node.prev
	h2 := node
	e2 := bc.Tail.prev
	linkNode(bc.Head, h2)
	linkNode(e2, h1)
	linkNode(e1, bc.Tail)
	bc._maintain()
}

func (bc *BlockChain[S, U, V]) Reverse(start, end int32) {
	if start >= end {
		return
	}
	end--
	left := bc._splitKth(start + 1)
	right := bc._splitKth(end + 2).prev
	begin := left.prev
	endNode := right.next
	right.next = nil
	bc._reverse(left, nil)
	linkNode(begin, right)
	linkNode(left, endNode)
	bc._maintain()
}

func (bc *BlockChain[S, U, V]) GetAll() []V {
	res := make([]V, 0, bc.size)
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		for i := int32(0); i < node.size; i++ {
			res = append(res, node.data.Get(i))
		}
	}
	return res
}

func (bc *BlockChain[S, U, V]) EnumerateBlock(f func(block IBlock[S, U, V])) {
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		f(node.data)
	}
}

func (bc *BlockChain[S, U, V]) Len() int32 {
	return bc.size
}

func (bc *BlockChain[S, U, V]) _reverse(root *linkedNode[IBlock[S, U, V]], p *linkedNode[IBlock[S, U, V]]) {
	if root == nil {
		return
	}
	bc._reverse(root.next, root)
	root.data.Reverse()
	root.prev = root.next
	root.next = p
}

func (bc *BlockChain[S, U, V]) _split(node *linkedNode[IBlock[S, U, V]], k int32) {
	post := &linkedNode[IBlock[S, U, V]]{}
	block1, block2 := node.data.Split(k)
	linkNode(post, node.next)
	linkNode(node, post)
	post.data = block2
	post.size = node.size - k
	node.data = block1
	node.size = k
}

func (bc *BlockChain[S, U, V]) _mergeNode(a, b *linkedNode[IBlock[S, U, V]]) {
	linkNode(a, b.next)
	a.data = a.data.Merge(b.data)
	a.size += b.size
}

func (bc *BlockChain[S, U, V]) _maintain() {
	bc.size = 0
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		bc.size += node.size
		if node.size >= 2*bc.b {
			bc._split(node, bc.b)
		} else if node.prev != bc.Head && node.size+node.prev.size <= bc.b {
			bc._mergeNode(node.prev, node)
		}
	}
}

func (bc *BlockChain[S, U, V]) _splitKth(k int32) *linkedNode[IBlock[S, U, V]] {
	for node := bc.Head.next; node != bc.Tail; node = node.next {
		if node.size < k {
			k -= node.size
			continue
		}
		if k != 1 {
			bc._split(node, k-1)
			node = node.next
		}
		return node
	}
	return bc.Tail
}

func (bc *BlockChain[S, U, V]) _enter(L, R, l, r int32) bool {
	return L <= l && r <= R
}

func (bc *BlockChain[S, U, V]) _leave(L, R, l, r int32) bool {
	return l > R || r < L
}

type linkedNode[V any] struct {
	prev, next *linkedNode[V]
	size       int32
	data       V
}

func (node *linkedNode[V]) String() string {
	return fmt.Sprint(node.data)
}

func linkNode[V any](a, b *linkedNode[V]) {
	b.prev = a
	a.next = b
}

func min32(a, b int32) int32 {
	if a < b {
		return a
	}
	return b
}

func max32(a, b int32) int32 {
	if a > b {
		return a
	}
	return b
}

// application
type INumberArrayBlock = IBlock[*int, int, int]
type NumberArrayBlock struct {
	data []int
	sum  int
}

func NewNumberArrayBlock(data []int) *NumberArrayBlock {
	res := &NumberArrayBlock{data: data}
	for _, v := range data {
		res.sum += v
	}
	return res
}

func (nab *NumberArrayBlock) Split(k int32) (block1 INumberArrayBlock, block2 INumberArrayBlock) {
	a, b := nab.data[:k], nab.data[k:]
	return NewNumberArrayBlock(a), NewNumberArrayBlock(b)
}

func (nab *NumberArrayBlock) Merge(other INumberArrayBlock) INumberArrayBlock {
	nab.data = append(nab.data, other.(*NumberArrayBlock).data...)
	for _, v := range other.(*NumberArrayBlock).data {
		nab.sum += v
	}
	return nab
}

func (nab *NumberArrayBlock) InserBefore(index int32, e int) {
	nab.data = append(nab.data[:index], append([]int{e}, nab.data[index:]...)...)
	nab.sum += e
}

func (nab *NumberArrayBlock) Delete(index int32) {
	nab.sum -= nab.data[index]
	nab.data = append(nab.data[:index], nab.data[index+1:]...)
}

func (nab *NumberArrayBlock) Get(index int32) int {
	return nab.data[index]
}

func (nab *NumberArrayBlock) Reverse() {
	for i, j := 0, len(nab.data)-1; i < j; i, j = i+1, j-1 {
		nab.data[i], nab.data[j] = nab.data[j], nab.data[i]
	}
}

func (nab *NumberArrayBlock) FullyQuery(sum *int) {
	*sum += nab.sum
}

func (nab *NumberArrayBlock) PartialQuery(index int32, sum *int) {
	*sum += nab.data[index]
}

func (nab *NumberArrayBlock) FullyUpdate(lazy int) {}

func (nab *NumberArrayBlock) PartialUpdate(index int32, lazy int) {}

func (nab *NumberArrayBlock) BeforePartialQuery() {}

func (nab *NumberArrayBlock) AfterPartialUpdate() {}

func test() {
	nums := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	blockChain := NewBlockChain(
		10,
		func(start, end int32) IBlock[*int, int, int] {
			return NewNumberArrayBlock(nums[start:end])
		},
		-1,
	)

	sum := new(int)
	blockChain.Query(0, 9, sum)
	fmt.Println(*sum)
}

// 406. 根据身高重建队列
// https://leetcode.cn/problems/queue-reconstruction-by-height/

File: 22_专题/离线查询/根号分治/块状链表/BlockChain.ts
// https://www.acwing.com/blog/content/28060/
// TODO: 查询聚合值有问题，谨慎使用

/**
 * S: 聚合类型.
 * U: 更新类型.
 * E: 元素类型.
 * B: 子类块类型.
 */
abstract class AbstractBlock<S, U, V, B extends AbstractBlock<S, U, V, B>> {
  /** 分裂整块，block1包含前k个元素，block2包含后面的元素.*/
  abstract split(k: number): { block1: B; block2: B }
  abstract merge(other: B): B
  /** 在index位置之前插入元素e，0表示第一个.*/
  abstract insertBefore(index: number, e: V): void
  abstract delete(index: number): void
  abstract get(index: number): V
  abstract reverse(): void

  abstract fullyQuery(sum: S): void
  abstract partialQuery(index: number, sum: S): void
  abstract fullyUpdate(lazy: U): void
  abstract partialUpdate(index: number, lazy: U): void
  beforePartialQuery(): void {}
  afterPartialUpdate(): void {}
}

class BlockChain<S, U, V, B extends AbstractBlock<S, U, V, B>> {
  /**
   * @param blockSize 块大小.默认为`2 * (1 + Math.sqrt(n))`.
   */
  static create<S, U, V, B extends AbstractBlock<S, U, V, B>>(
    n: number,
    blockSupplier: (start: number, end: number) => B,
    blockSize = 2 * (1 + (Math.sqrt(n) | 0))
  ): BlockChain<S, U, V, B> {
    const res = new BlockChain<S, U, V, B>()
    res._b = blockSize
    res._size = n
    LinkedNode.link(res.head, res.tail)
    for (let start = 0; start < n; start += blockSize) {
      let end = start + blockSize
      if (end > n) end = n
      const block = blockSupplier(start, end)
      const node = new LinkedNode<B>()
      node.data = block
      node.size = end - start
      LinkedNode.link(res.tail.prev, node)
      LinkedNode.link(node, res.tail)
    }
    return res
  }

  private static new2<S, U, V, B extends AbstractBlock<S, U, V, B>>(
    b: number,
    supplier: () => B
  ): BlockChain<S, U, V, B> {
    const res = new BlockChain<S, U, V, B>()
    const block = supplier()
    const node = new LinkedNode<B>()
    node.data = block
    node.size = 0
    res._b = b
    LinkedNode.link(res.tail.prev, node)
    LinkedNode.link(node, res.tail)
    return res
  }

  private static new3<S, U, V, B extends AbstractBlock<S, U, V, B>>(
    b: number,
    begin: LinkedNode<B>,
    end: LinkedNode<B>
  ): BlockChain<S, U, V, B> {
    // add an empty node
    const res = new BlockChain<S, U, V, B>()
    res._b = b
    LinkedNode.link(res.head, begin)
    LinkedNode.link(end, res.tail)
    res._maintain()
    return res
  }

  head = new LinkedNode<B>()
  tail = new LinkedNode<B>()
  private _b = 0
  private _size = 0

  private constructor() {}

  get(index: number): V {
    if (index < 0) index += this._size
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      if (node.size <= index) {
        index -= node.size
        continue
      }
      node.data!.beforePartialQuery()
      return node.data!.get(index)
    }
    throw new Error('Index out of bounds')
  }

  prefixSize(block: B, include: boolean): number {
    let res = 0
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      if (node.data === block) {
        if (include) {
          res += node.size
        }
        break
      }
      res += node.size
    }
    return res
  }

  split(
    k: number,
    blockSupplier: () => B
  ): { first: BlockChain<S, U, V, B>; second: BlockChain<S, U, V, B> } {
    k++
    if (k === 1) {
      return { first: BlockChain.new2(this._b, blockSupplier), second: this }
    }
    if (k > this._size) {
      return { first: this, second: BlockChain.new2(this._b, blockSupplier) }
    }
    const head = this._splitKth(k)
    const end = this.tail.prev
    LinkedNode.link(head.prev, this.tail)
    const b = BlockChain.new3<S, U, V, B>(this._b, head, end)
    this._maintain()
    return { first: this, second: b }
  }

  mergeDestructively(other: BlockChain<S, U, V, B>): BlockChain<S, U, V, B> {
    LinkedNode.link(this.tail.prev, other.head.next)
    this.tail = other.tail
    this._size += other._size
    return this
  }

  insertBefore(index: number, e: V): void {
    if (index < 0) index += this._size
    if (index < 0) index = 0
    if (index > this._size) index = this._size
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      if (node.size < index) {
        index -= node.size
        continue
      }
      node.data!.insertBefore(index, e)
      node.size++
      break
    }
    this._maintain()
  }

  delete(index: number): void {
    if (index < 0) index += this._size
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      if (node.size <= index) {
        index -= node.size
        continue
      }
      node.data!.delete(index)
      node.size--
      break
    }
    this._maintain()
  }

  update(start: number, end: number, update: U): void {
    if (start < 0) start += this._size
    if (end > this._size) end = this._size
    if (start >= end) return
    end--
    let offset = 0
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      const left = offset
      const right = offset + node.size - 1
      offset += node.size
      if (this._enter(start, end, left, right)) {
        node.data!.fullyUpdate(update)
      } else if (this._leave(start, end, left, right)) {
        continue
      } else {
        for (let i = Math.max(left, start), to = Math.min(right, end); i <= to; i++) {
          node.data!.partialUpdate(i - left, update)
        }
        node.data!.afterPartialUpdate()
      }
    }
  }

  query(start: number, end: number, sum: S): void {
    if (start < 0) start += this._size
    if (end > this._size) end = this._size
    if (start >= end) return
    end--
    let offset = 0
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      const left = offset
      const right = offset + node.size - 1
      offset += node.size
      if (this._enter(start, end, left, right)) {
        node.data!.fullyQuery(sum)
      } else if (this._leave(start, end, left, right)) {
        continue
      } else {
        node.data!.beforePartialQuery()
        for (let i = Math.max(left, start), to = Math.min(right, end); i <= to; i++) {
          node.data!.partialQuery(i - left, sum)
        }
      }
    }
  }

  /** 向左旋转k个位置.*/
  rotateLeft(k: number): void {
    if (k < 0) k += this._size
    if (k >= this._size) k %= this._size
    if (k === 0) return
    k++
    const node = this._splitKth(k)
    const h1 = this.head.next
    const e1 = node.prev
    const h2 = node
    const e2 = this.tail.prev
    LinkedNode.link(this.head, h2)
    LinkedNode.link(e2, h1)
    LinkedNode.link(e1, this.tail)
    this._maintain()
  }

  reverse(start = 0, end = this._size): void {
    if (start >= end) return
    end--
    const left = this._splitKth(start + 1)
    const right = this._splitKth(end + 2).prev
    const begin = left.prev
    const endNode = right.next
    right.next = LinkedNode._NIL
    this._reverse(left, LinkedNode._NIL)
    LinkedNode.link(begin, right)
    LinkedNode.link(left, endNode)
    this._maintain()
  }

  getAll(): V[] {
    const res: V[] = []
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      for (let i = 0; i < node.size; i++) {
        res.push(node.data!.get(i))
      }
    }
    return res
  }

  enumerateNode(f: (node: LinkedNode<B>) => void): void {
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      f(node)
    }
  }

  get length(): number {
    return this._size
  }

  private _reverse(root: LinkedNode<B>, p: LinkedNode<B>): void {
    if (root === LinkedNode._NIL) {
      return
    }
    this._reverse(root.next, root)
    root.data!.reverse()
    root.prev = root.next
    root.next = p
  }

  private _split(node: LinkedNode<B>, k: number): void {
    const post = new LinkedNode<B>()
    const { block1, block2 } = node.data!.split(k)
    LinkedNode.link(post, node.next)
    LinkedNode.link(node, post)
    post.data = block2
    post.size = node.size - k
    node.data = block1
    node.size = k
  }

  private _mergeNode(a: LinkedNode<B>, b: LinkedNode<B>): void {
    LinkedNode.link(a, b.next)
    a.data = a.data!.merge(b.data!)
    a.size += b.size
  }

  private _maintain(): void {
    this._size = 0
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      this._size += node.size
      if (node.size >= 2 * this._b) {
        this._split(node, this._b)
      } else if (node.prev !== this.head && node.size + node.prev.size <= this._b) {
        this._mergeNode(node.prev, node)
      }
    }
  }

  /**
   * 拆分成前k个元素和后面的元素，返回后面的元素.
   */
  private _splitKth(k: number): LinkedNode<B> {
    for (let node = this.head.next; node !== this.tail; node = node.next) {
      if (node.size < k) {
        k -= node.size
        continue
      }
      if (k !== 1) {
        this._split(node, k - 1)
        node = node.next
      }
      return node
    }
    return this.tail
  }

  private _enter(L: number, R: number, l: number, r: number): boolean {
    return L <= l && r <= R
  }

  private _leave(L: number, R: number, l: number, r: number): boolean {
    return l > R || r < L
  }
}

class LinkedNode<E> {
  static readonly _NIL = new LinkedNode<any>()

  static link<E>(a: LinkedNode<E>, b: LinkedNode<E>): void {
    b.prev = a
    a.next = b
  }

  prev: LinkedNode<E> = LinkedNode._NIL
  next: LinkedNode<E> = LinkedNode._NIL
  data: E | undefined = undefined
  size = 0
}

export { BlockChain, AbstractBlock }
